test/celeritas/io/EventIO.test.cc

```diff
@@ -0,0 +1,260 @@
+        {0, 0, 1},
+        {2.326451417389850e-2, -4.866936365179566e-2, 9.985439676959555e-1},
+        {0, 0, -1},
+        {-5.260794237813896e-2, -3.280442747570201e-1, -9.431963518790131e-1},
+        {-9.009470900796461e-1, 2.669997932835038e-2, -4.331067443262500e-1},
+        {5.189457940206315e-2, -7.074356638330033e-1, -7.048700122475354e-1},
+        {-8.273504806466310e-2, 9.750892208717103e-1, 2.058055469649411e-1},
+        {7.028153760960004e-2, -8.780402697122620e-1, -4.733981307893478e-1}};
+
+    // Read events from the event record
+    int event_count = 0;
+    std::vector<Primary> primaries;
+    do
+    {
+        primaries = read_event();
+        ASSERT_TRUE(primaries.empty()
+                    || primaries.size() == std::size(expected_pdg));
+        for (auto i : range(primaries.size()))
+        {
+            auto const& primary = primaries[i];
+
+            // Check that the particle types were read correctly
+            EXPECT_EQ(
+                expected_pdg[i],
+                particles_->id_to_pdg(primary.particle_id).unchecked_get());
+
+            // Check that the event IDs match
+            EXPECT_EQ(event_count, primary.event_id.unchecked_get());
+
+            // Check that the position, direction, and energy were read
+            // correctly
+            double const expected_position[] = {0, 0, 0};
+            EXPECT_VEC_SOFT_EQ(expected_position, primary.position);
+            EXPECT_VEC_SOFT_EQ(expected_direction[i], primary.direction);
+            EXPECT_DOUBLE_EQ(expected_energy[i],
+                             value_as<MevEnergy>(primary.energy));
+            EXPECT_EQ(0, primary.time);
+        }
+        ++event_count;
+    } while (!primaries.empty());
+    // Event reader should keep returning an empty vector
+    primaries = read_event();
+    EXPECT_TRUE(primaries.empty());
+}
+
+TEST_P(EventIO, write_read)
+{
+    std::vector<Primary> primaries = [&particles = *this->particles_] {
+        auto proton_id = particles.find(pdg::proton());
+        auto gamma_id = particles.find(pdg::gamma());
+        Primary gamma{gamma_id,
+                      MevEnergy{1.23},
+                      Real3{2, 4, 5},
+                      Real3{1, 0, 0},
+                      5.67e-9 * units::second,
+                      EventId{0},
+                      TrackId{}};
+        Primary proton{proton_id,
+                       MevEnergy{2.34},
+                       Real3{3, 5, 8},
+                       Real3{0, 1, 0},
+                       5.78e-9 * units::second,
+                       EventId{0},
+                       TrackId{}};
+        std::vector<Primary> primaries{gamma, proton, gamma, proton};
+        primaries[1].position = {-3, -4, 5};
+        primaries[3].position = primaries[2].position;
+        primaries[3].time = primaries[2].time;
+        for (auto i : range(primaries.size()))
+        {
+            primaries[i].track_id = TrackId{i * 2};
```

Comment:
```suggestion
            primaries[i].track_id = TrackId(i * 2);
```
(error: narrowing conversion of ‘(i * 2)’...)